src/utils/hotcode-agent/src/main/java/io/bellsoft/hotcode/agent/HotCodeMain.java

```diff
@@ -0,0 +1,153 @@
+            attach(pid, keyArgs);
+            break;
+        default:
+            usage();
+            break;
+        }
+    }
+
+    private static void attach(String pid, Properties keyArgs) {
+        var argumentString = agentArgs(keyArgs);
+        var jar = HotCodeMain.class.getProtectionDomain().getCodeSource().getLocation().getPath();
+        try {
+            var vm = VirtualMachine.attach(pid);
+            vm.loadAgent(jar, argumentString);
+            vm.detach();
+        } catch (Exception e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    private static void convert(Properties keyArgs, Path recordingPath, Path directiveFilePath) {
+        var config = HotCodeAgentConfiguration.from(keyArgs);
+        var profile = new JfrOfflineProfiling(config.top, config.maxStackDepth, recordingPath).call();
+
+        new MethodProfilePrinter(System.out).print(profile);
+
+        var hotMethods = profile.getTop();
+        var directives = CompilerDirectives.build(hotMethods);
+        
+        try {
+            Files.writeString(directiveFilePath, directives);
+        } catch (IOException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    private static void parseCommandArgs(String[] args, Properties keyArgs, ArrayList<String> freeArgs) {
+        String key = null;
+        
+        for (int i = 1; i < args.length; i++) {
+            if (args[i].startsWith("--")) {
+                // check for key without value
+                if (key != null) {
+                    usage();
+                }
+                key = args[i].substring(2);
+                if (key.isEmpty()) {
+                    usage();
+                }
+            } else {
+                var value = args[i];
+                if (key != null) {
+                    keyArgs.put(key, value);
+                    key = null;
+                } else {
+                    freeArgs.add(value);
+                }
+            }
+        }
+        // check for key without value
+        if (key != null) {
+            usage();
+        }
+    }
+    
+    private static String agentArgs(Properties keyArgs) {
+        var sb = new StringBuilder();
+        for (var key : keyArgs.keySet()) {
+            if (!sb.isEmpty()) {
+                sb.append(",");
+            }
+            sb.append(key).append("=").append(keyArgs.get(key));
+        }
+        return sb.toString();
+    }
+
+    private static void checkBasicCommands(String command) {
+        switch (command) {
+        case "-h":
+        case "-?":
+        case "--help":
+        case "help":
+            usage();
+            break;
+        case "-v":
+        case "--version":
+        case "version":
+            version();
+            break;
+        }
+    }
+ 
+    static void usage() {
+        try (var in = HotCodeMain.class.getResourceAsStream(USAGE_TXT)) {
+            System.out.println(new String(in.readAllBytes()));
```

Comment:
Unnecessary conversion to String: `System.out.write(in.readAllBytes())`

Reply:
...or transferTo. But here we can easily support different encodings of source and target - not a realistic case but it is about the idea of writing messages not bytes to stdout.